crates/gpui/src/platform/linux/text_system.rs

```diff
@@ -55,42 +60,118 @@ impl Default for LinuxTextSystem {
 #[allow(unused)]
 impl PlatformTextSystem for LinuxTextSystem {
     fn add_fonts(&self, fonts: Vec<Cow<'static, [u8]>>) -> Result<()> {
-        Ok(()) //TODO
+        self.0.write().add_fonts(fonts)
     }
+
     fn all_font_names(&self) -> Vec<String> {
-        Vec::new()
+        self.0.read().font_system.db().faces().map(|face| face.post_script_name.clone()).collect()
     }
+
     fn all_font_families(&self) -> Vec<String> {
         Vec::new()
     }
-    fn font_id(&self, descriptor: &Font) -> Result<FontId> {
-        Ok(FontId(0)) //TODO
+
+    fn font_id(&self, font: &Font) -> Result<FontId> {
+        let lock = self.0.upgradable_read();
+        if let Some(font_id) = lock.font_selections.get(font) {
+            Ok(*font_id)
+        } else {
+            let mut lock = RwLockUpgradableReadGuard::upgrade(lock);
+            let candidates = if let Some(font_ids) = lock.font_ids_by_family_name.get(&font.family)
+            {
+                font_ids.as_slice()
+            } else {
+                let font_ids = lock.load_family(&font.family, font.features)?;
+                lock.font_ids_by_family_name
+                    .insert(font.family.clone(), font_ids);
+                lock.font_ids_by_family_name[&font.family].as_ref()
+            };
+
+            let id = lock
+                .font_system
+                .db()
+                .query(&Query {
+                    families: &[Family::Name(&font.family)],
+                    weight: font.weight.into(),
+                    style: font.style.into(),
+                    stretch: Default::default(),
+                })
+                .context("no font")?;
+            // println!("{:?}", id);
+            // println!("{:?}", lock.fonts);
+            let font_id = if let Some(font_id) = lock.fonts.iter().position(|font| font.id() == id)
+            {
+                FontId(font_id)
+            } else {
+                // HACK: font isn't in fonts so add it there, this is because we query all the fonts in the db and maybe we haven't loaded it yet
+                let font_id = FontId(lock.fonts.len());
+                let font = lock.font_system.get_font(id).unwrap();
+                lock.fonts.push(font);
+                font_id
+            };
+
+            lock.font_selections.insert(font.clone(), font_id);
+            Ok(font_id)
+        }
     }
+
     fn font_metrics(&self, font_id: FontId) -> FontMetrics {
-        unimplemented!()
+        let metrics = self.0.read().fonts[font_id.0].as_swash().metrics(&[]);
+        FontMetrics {
+            units_per_em: metrics.units_per_em as u32,
+            ascent: metrics.ascent,
+            descent: metrics.descent - 1000.0, // HACK: extremly hacky but this means text doesn't get cut off, something to do with gutter padding in the editor
```

Comment:
instead of subtracting 1000 I think this should just be ``-metrics.descent`` like in https://github.com/pop-os/cosmic-text/blob/990d66ed41df4fae5db23ec10c9d6c0e95854449/src/shape.rs#L103